src/esp/gfx_batch/Hbao.h

```diff
@@ -0,0 +1,180 @@
+#ifndef ESP_GFX_BATCH_HBAO_H_
+#define ESP_GFX_BATCH_HBAO_H_
+
+#include <Corrade/Containers/Pointer.h>
+#include <Magnum/GL/GL.h>
+#include <Magnum/Magnum.h>
+#include <Magnum/Math/Vector2.h>
+
+namespace esp {
+namespace gfx_batch {
+
+enum class HbaoFlag {
+  Blur = 1 << 0,
+
+  /**
+   * Special blur is dependent on blur being set, and is ignored if blur is not
+   * set.
+   */
+  UseAoSpecialBlur = 1 << 1,
+  /* These two affect only the cache-aware variant. Only one can be set at
+     most, not both (mutually exclusive) */
+  LayeredImageLoadStore = 1 << 2,
+  LayeredGeometryShader = 1 << 3
+};
+typedef Corrade::Containers::EnumSet<HbaoFlag> HbaoFlags;
+
+CORRADE_ENUMSET_OPERATORS(HbaoFlags)
+
+class HbaoConfiguration {
+ public:
+  Magnum::Vector2i size() const { return size_; }
+
+  HbaoConfiguration& setSize(const Magnum::Vector2i& size) {
+    size_ = size;
+    return *this;
+  }
+
+  HbaoFlags flags() const { return flags_; }
+
+  HbaoConfiguration& setFlags(HbaoFlags flags) {
+    flags_ = flags;
+    return *this;
+  }
+
+  // Blur should always be used, otherwise there's nasty grid-like artifacts, at
+  // least until we support MSAA
+
+  // HbaoConfiguration& setUseBlur(bool state) {
+  //   // special blur is dependent on blur being enabled, so both should be
+  //   // disabled if blur is.
+  //   flags_ = (state ? flags_ | HbaoFlag::Blur
+  //                   : flags_ & ~(HbaoFlag::UseAoSpecialBlur |
+  //                   HbaoFlag::Blur));
+  //   return *this;
+  // }
+
+  HbaoConfiguration& setUseSpecialBlur(bool state) {
+    // special blur is dependent on blur being enabled, so both should be
+    // enabled if special blur is
+    flags_ = (state ? flags_ | (HbaoFlag::UseAoSpecialBlur | HbaoFlag::Blur)
+                    : flags_ & ~HbaoFlag::UseAoSpecialBlur);
+    return *this;
+  }
+
+  HbaoConfiguration& setUseLayeredImageLoadStore(bool state) {
+    // LayeredImageLoadStore and LayeredGeometryShader are mutually exclusive
+    flags_ = (state ? (flags_ | HbaoFlag::LayeredImageLoadStore) &
+                          ~HbaoFlag::LayeredGeometryShader
+                    : flags_ & ~HbaoFlag::LayeredImageLoadStore);
+    return *this;
+  }
+  HbaoConfiguration& setUseLayeredGeometryShader(bool state) {
+    // LayeredImageLoadStore and LayeredGeometryShader are mutually exclusive
+    flags_ = (state ? (flags_ | HbaoFlag::LayeredGeometryShader) &
+                          ~HbaoFlag::LayeredImageLoadStore
+                    : flags_ & ~HbaoFlag::LayeredGeometryShader);
+    return *this;
+  }
```

Comment:
I don't know, wouldn't it be easier to just make the flags supersets of one another? I didn't do that originally because it required extra brain cycles, but

```cpp
enum class HbaoFlag {
  Blur = 1 << 0,
  UseAoSpecialBlur = (1 << 1)|Blur,
  LayeredImageLoadStore = 1 << 2,
  LayeredGeometryShader = (1 << 3)|LayeredImageLoadStore
};
```

should cover pretty much all these cases without needing extra setters. I.e., if you enable `UseAoSpecialBlur`, it'll add `Blur` as well, if you enable `LayeredImageLoadStore|LayeredGeometryShader` it'll pick the GS as "higher priority" (assuming the passthrough is faster than image load/store). So it's then just a matter of documenting the relations.

With this you of course need to test for the flags with `>=` instead of `&` in the code, but that's a simple change.

Reply:
I'm going to change these as we discussed, with a "no blur" flag that turns off all blur, and the UseAOSpecialBlur flag to use the special blur, with the non-special blur being the default behavior.